receiver/k8sclusterreceiver/collection/metadata.go

```diff
@@ -25,42 +25,158 @@ import (
+			out[id] = km
+		}
+	}
+
+	return out
+}
+
+// KubernetesMetadataExporter provides an interface to implement
+// ConsumeKubernetesMetadata in Exporters that support metadata.
+type KubernetesMetadataExporter interface {
+	// ConsumeKubernetesMetadata will be invoked every time there's an
+	// update to a resource that results in one or more KubernetesMetadataUpdate.
+	ConsumeKubernetesMetadata(metadata []*KubernetesMetadataUpdate) error
+}
+
+// KubernetesMetadataUpdate provides a delta view of metadata on a resource between
+// two revisions of a resource.
+type KubernetesMetadataUpdate struct {
+	// ResourceIDKey is the label key of UID label for the resource.
+	ResourceIDKey string
+	// ResourceID is the Kubernetes UID of the resource. In case of
+	// containers, this value is the container id.
+	ResourceID string
+	// MetadataToAdd contains key-value pairs that are newly added to
+	// the resource description in the current revision.
+	MetadataToAdd map[string]string
+	// MetadataToUpdate contains key-value pairs that have been updated
+	// in the current revision compared to the previous revisions(s).
+	MetadataToUpdate map[string]string
+	// MetadataToRemove contains key-value pairs that no longer describe
+	// a resource and needs to be removed.
+	MetadataToRemove map[string]string
+}
+
+// GetKubernetesMetadataUpdate processes kubernetes metadata updates and returns
+// a map of a delta of metadata mapped to each resource.
+func GetKubernetesMetadataUpdate(old, new map[string]*KubernetesMetadata) []*KubernetesMetadataUpdate {
+	var out []*KubernetesMetadataUpdate
+
+	for id, oldMetadata := range old {
+		// if an object with the same id has a previous revision, take a delta
+		// of the metadata.
+		if newMetadata, ok := new[id]; ok {
+			if toAdd, toRemove, toUpdate, isNonEmpty :=
+				getPropertiesDelta(oldMetadata.metadata, newMetadata.metadata); isNonEmpty {
+				out = append(out, &KubernetesMetadataUpdate{
+					ResourceIDKey:    oldMetadata.resourceIDKey,
+					ResourceID:       id,
+					MetadataToAdd:    toAdd,
+					MetadataToRemove: toRemove,
+					MetadataToUpdate: toUpdate,
+				})
+			}
+		}
+	}
+
+	// In case there are resources in the current revision, that was not in the
+	// previous revision, collect metadata to be added
+	for id, km := range new {
+		// if an id is seen for the first time, all metadata need to be added.
+		if _, ok := old[id]; !ok {
+			out = append(out, &KubernetesMetadataUpdate{
+				ResourceIDKey: km.resourceIDKey,
+				ResourceID:    id,
+				MetadataToAdd: km.metadata,
+			})
+		}
+	}
+
+	return out
+}
+
+// getPropertiesDelta returns 3 maps one each representing properties to be added,
+// removed and updated. It also returns a boolean as the forth return value. This
+// value is true if at least one of the 3 maps are non empty, otherwise false.
+func getPropertiesDelta(oldProps,
+	newProps map[string]string) (map[string]string, map[string]string, map[string]string, bool) {
```

Comment:
3 maps of the same type in the return are very easy to confuse the order. It will be nicer to declare a struct with named fields and return the struct by value instead.

Reply:
Wrapped these maps around `MetadataDelta`.